File: packages/python-sdk/src/treedx_sdk/adapters/files.py
from __future__ import annotations

from typing import Any, Mapping

from .common import json_request, segment
from treedx_sdk.transport import Transport


class FilesAdapter:
    def __init__(self, transport: Transport) -> None:
        self.transport = transport

    def tree(self, workspace_id: str, query: Mapping[str, str | int | float | bool | None] | None = None) -> Any:
        return json_request(self.transport, "GET", f"/api/v1/workspaces/{segment(workspace_id)}/tree", query=query)

    def read(self, workspace_id: str, query: Mapping[str, str | int | float | bool | None] | None = None) -> Any:
        return json_request(self.transport, "GET", f"/api/v1/workspaces/{segment(workspace_id)}/files", query=query)

    def write(self, workspace_id: str, body: Any) -> Any:
        return json_request(self.transport, "PUT", f"/api/v1/workspaces/{segment(workspace_id)}/files", body)

    def patch(self, workspace_id: str, body: Any) -> Any:
        return json_request(self.transport, "PATCH", f"/api/v1/workspaces/{segment(workspace_id)}/files", body)

    def delete(self, workspace_id: str, query: Mapping[str, str | int | float | bool | None] | None = None) -> Any:
        return json_request(self.transport, "DELETE", f"/api/v1/workspaces/{segment(workspace_id)}/files", query=query)

    def search(self, workspace_id: str, body: Any) -> Any:
        return json_request(self.transport, "POST", f"/api/v1/workspaces/{segment(workspace_id)}/search", body)

    def status(self, workspace_id: str) -> Any:
        return json_request(self.transport, "GET", f"/api/v1/workspaces/{segment(workspace_id)}/status")

    def diff(self, workspace_id: str, query: Mapping[str, str | int | float | bool | None] | None = None) -> Any:
        return json_request(self.transport, "GET", f"/api/v1/workspaces/{segment(workspace_id)}/diff", query=query)

    def commit(self, workspace_id: str, body: Any) -> Any:
        return json_request(self.transport, "POST", f"/api/v1/workspaces/{segment(workspace_id)}/commit", body)

File: packages/elixir-sdk/lib/treedx_sdk/adapters/files.ex
defmodule TreeDxSdk.Files do
  @moduledoc false
  alias TreeDxSdk.Adapters.Common

  def tree(client, workspace_id, query \\ %{}) do
    Common.json_request(
      client,
      :get,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/tree",
      nil,
      query
    )
  end

  def read(client, workspace_id, query \\ %{}) do
    Common.json_request(
      client,
      :get,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/files",
      nil,
      query
    )
  end

  def write(client, workspace_id, body) do
    Common.json_request(
      client,
      :put,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/files",
      body,
      %{}
    )
  end

  def patch(client, workspace_id, body) do
    Common.json_request(
      client,
      :patch,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/files",
      body,
      %{}
    )
  end

  def delete(client, workspace_id, query \\ %{}) do
    Common.json_request(
      client,
      :delete,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/files",
      nil,
      query
    )
  end

  def search(client, workspace_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/search",
      body,
      %{}
    )
  end

  def status(client, workspace_id) do
    Common.json_request(
      client,
      :get,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/status",
      nil,
      %{}
    )
  end

  def diff(client, workspace_id, query \\ %{}) do
    Common.json_request(
      client,
      :get,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/diff",
      nil,
      query
    )
  end

  def commit(client, workspace_id, body) do
    Common.json_request(
      client,
      :post,
      "/api/v1/workspaces/" <> Common.segment(workspace_id) <> "/commit",
      body,
      %{}
    )
  end
end
